File: EBiletJava/app/src/main/java/com/alperensarac/ebiletjava/ui/home/HomeActivity.java
package com.alperensarac.ebiletjava.ui.home;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.alperensarac.ebiletjava.R;
import com.alperensarac.ebiletjava.data.api.ApiClient;
import com.alperensarac.ebiletjava.data.model.ApiResponse;
import com.alperensarac.ebiletjava.data.model.City;
import com.alperensarac.ebiletjava.data.model.District;
import com.alperensarac.ebiletjava.data.model.Event;
import com.alperensarac.ebiletjava.data.session.SessionManager;
import com.alperensarac.ebiletjava.ui.auth.LoginActivity;
import com.alperensarac.ebiletjava.ui.event.EventDetailActivity;
import com.alperensarac.ebiletjava.ui.scanner.TicketScannerActivity;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class HomeActivity extends AppCompatActivity {

    /*
        Üst bar view'ları.
    */
    private TextView tvWelcome;
    private TextView tvRoleInfo;
    private Button btnMyTickets;
    private Button btnLogout;
    private Button btnScanner;

    /*
        Filtre view'ları.
    */
    private Spinner spinnerCities;
    private Spinner spinnerDistricts;
    private Button btnListEvents;
    private TextView tvStatus;

    /*
        RecyclerView.
    */
    private RecyclerView rvEvents;
    private EventAdapter eventAdapter;

    /*
        Session bilgileri.
    */
    private SessionManager sessionManager;

    /*
        Spinner için şehir ve ilçe listeleri.
    */
    private final List<City> cityList = new ArrayList<>();
    private final List<District> districtList = new ArrayList<>();

    /*
        Seçili şehir ve ilçe.
    */
    private City selectedCity = null;
    private District selectedDistrict = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        sessionManager = new SessionManager(this);

        /*
            Bu ekran token olmadan çalışmamalı.
        */
        if (!sessionManager.isLoggedIn()) {
            goToLogin();
            return;
        }

        initViews();
        setupHeader();
        setupRecyclerView();
        setupClickListeners();

        /*
            Ekran açılır açılmaz şehirleri backend'den çekiyoruz.
        */
        loadCities();
    }

    /*
        XML view'larını Java değişkenlerine bağlar.
    */
    private void initViews() {
        tvWelcome = findViewById(R.id.tvWelcome);
        tvRoleInfo = findViewById(R.id.tvRoleInfo);
        btnMyTickets = findViewById(R.id.btnMyTickets);
        btnLogout = findViewById(R.id.btnLogout);
        btnScanner = findViewById(R.id.btnScanner);

        spinnerCities = findViewById(R.id.spinnerCities);
        spinnerDistricts = findViewById(R.id.spinnerDistricts);
        btnListEvents = findViewById(R.id.btnListEvents);
        tvStatus = findViewById(R.id.tvStatus);

        rvEvents = findViewById(R.id.rvEvents);
    }

    /*
        Üst bar kullanıcı bilgilerini doldurur.
    */
    private void setupHeader() {
        tvWelcome.setText("Hoş geldin, " + sessionManager.getFullName());

        String role = sessionManager.getRole();
        String roleText;

        if ("admin".equals(role)) {
            roleText = "Admin hesabı";
        } else if ("staff".equals(role)) {
            roleText = "Görevli hesabı";
        } else {
            roleText = "Etkinlikleri keşfet";
        }

        tvRoleInfo.setText(roleText);

    /*
        QR Kontrol butonu sadece staff ve admin hesaplarda görünür.
    */
        if (sessionManager.isStaffOrAdmin()) {
            btnScanner.setVisibility(View.VISIBLE);
        } else {
            btnScanner.setVisibility(View.GONE);
        }
    }

    /*
        RecyclerView kurulumunu yapar.
    */
    private void setupRecyclerView() {
        eventAdapter = new EventAdapter(new EventAdapter.OnEventClickListener() {
            @Override
            public void onEventClick(Event event) {
            /*
                Etkinlik kartına tıklanınca detay ekranına geçiyoruz.

                Burada sadece event_id gönderiyoruz.
                Detay ekranı geri kalan bilgileri backend'den çekecek.
            */
                Intent intent = new Intent(HomeActivity.this, EventDetailActivity.class);
                intent.putExtra(EventDetailActivity.EXTRA_EVENT_ID, event.getId());
                startActivity(intent);
            }
        });

        rvEvents.setLayoutManager(new LinearLayoutManager(this));
        rvEvents.setAdapter(eventAdapter);
    }

    /*
        Buton ve spinner olaylarını ayarlar.
    */
    private void setupClickListeners() {

        /*
            Biletlerim ekranı henüz yok.
            Şimdilik Toast gösteriyoruz.
        */
        btnMyTickets.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Toast.makeText(
                        HomeActivity.this,
                        "Biletlerim ekranı sonraki adımda eklenecek",
                        Toast.LENGTH_SHORT
                ).show();
            }
        });
        btnScanner.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(HomeActivity.this, TicketScannerActivity.class);
                startActivity(intent);
            }
        });

        /*
            Çıkış yapma.
        */
        btnLogout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sessionManager.logout();
                goToLogin();
            }
        });

        /*
            Şehir seçimi.

            Kullanıcı şehir seçince:
            - selectedCity güncellenir
            - eski ilçeler temizlenir
            - eski etkinlikler temizlenir
            - seçilen şehre ait ilçeler backend'den çekilir
        */
        spinnerCities.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(
                    AdapterView<?> parent,
                    View view,
                    int position,
                    long id
            ) {
                if (cityList.isEmpty()) {
                    return;
                }

                selectedCity = cityList.get(position);

                selectedDistrict = null;
                districtList.clear();
                eventAdapter.updateList(new ArrayList<Event>());

                loadDistricts(selectedCity.getId());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                selectedCity = null;
            }
        });

        /*
            İlçe seçimi.
        */
        spinnerDistricts.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(
                    AdapterView<?> parent,
                    View view,
                    int position,
                    long id
            ) {
                if (districtList.isEmpty()) {
                    return;
                }

                selectedDistrict = districtList.get(position);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                selectedDistrict = null;
            }
        });

        /*
            Etkinlikleri listele.
        */
        btnListEvents.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (selectedCity == null) {
                    Toast.makeText(
                            HomeActivity.this,
                            "Lütfen şehir seçiniz",
                            Toast.LENGTH_SHORT
                    ).show();
                    return;
                }

                if (selectedDistrict == null) {
                    Toast.makeText(
                            HomeActivity.this,
                            "Lütfen ilçe seçiniz",
                            Toast.LENGTH_SHORT
                    ).show();
                    return;
                }

                loadEvents(selectedCity.getId(), selectedDistrict.getId());
            }
        });
    }

    /*
        Şehirleri backend'den çeker.

        API:
        locations/cities_list.php

        POST:
        api_token
    */
    private void loadCities() {
        setStatus("Şehirler yükleniyor...");

        String apiToken = sessionManager.getApiToken();

        ApiClient.getApiService()
                .getCities(apiToken)
                .enqueue(new Callback<ApiResponse<List<City>>>() {
                    @Override
                    public void onResponse(
                            Call<ApiResponse<List<City>>> call,
                            Response<ApiResponse<List<City>>> response
                    ) {
                        if (!response.isSuccessful()) {
                            setStatus("Şehirler alınamadı. Sunucu hatası: " + response.code());
                            return;
                        }

                        ApiResponse<List<City>> apiResponse = response.body();

                        if (apiResponse == null) {
                            setStatus("Şehirler alınamadı. Boş cevap döndü.");
                            return;
                        }

                        if (!apiResponse.isSuccess()) {
                            setStatus(apiResponse.getMessage());
                            return;
                        }

                        List<City> cities = apiResponse.getData();

                        cityList.clear();

                        if (cities != null) {
                            cityList.addAll(cities);
                        }

                        setupCitySpinner();

                        if (cityList.isEmpty()) {
                            setStatus("Aktif şehir bulunamadı.");
                        } else {
                            setStatus("Şehir seçiniz.");
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<List<City>>> call, Throwable t) {
                        setStatus("Bağlantı hatası: " + t.getLocalizedMessage());
                    }
                });
    }

    /*
        Şehir Spinner'ını doldurur.

        City.java içinde toString() name döndürdüğü için
        Spinner ekranda şehir adını gösterir.
    */
    private void setupCitySpinner() {
        ArrayAdapter<City> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                cityList
        );

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        spinnerCities.setAdapter(adapter);
    }

    /*
        Seçilen şehre göre ilçeleri backend'den çeker.

        API:
        locations/districts_by_city.php

        POST:
        api_token
        city_id
    */
    private void loadDistricts(int cityId) {
        setStatus("İlçeler yükleniyor...");

        String apiToken = sessionManager.getApiToken();

        ApiClient.getApiService()
                .getDistrictsByCity(apiToken, cityId)
                .enqueue(new Callback<ApiResponse<List<District>>>() {
                    @Override
                    public void onResponse(
                            Call<ApiResponse<List<District>>> call,
                            Response<ApiResponse<List<District>>> response
                    ) {
                        if (!response.isSuccessful()) {
                            setStatus("İlçeler alınamadı. Sunucu hatası: " + response.code());
                            return;
                        }

                        ApiResponse<List<District>> apiResponse = response.body();

                        if (apiResponse == null) {
                            setStatus("İlçeler alınamadı. Boş cevap döndü.");
                            return;
                        }

                        if (!apiResponse.isSuccess()) {
                            setStatus(apiResponse.getMessage());
                            return;
                        }

                        List<District> districts = apiResponse.getData();

                        districtList.clear();

                        if (districts != null) {
                            districtList.addAll(districts);
                        }

                        setupDistrictSpinner();

                        if (districtList.isEmpty()) {
                            setStatus("Bu şehir için aktif ilçe bulunamadı.");
                        } else {
                            setStatus("İlçe seçip etkinlikleri listeleyebilirsin.");
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<List<District>>> call, Throwable t) {
                        setStatus("Bağlantı hatası: " + t.getLocalizedMessage());
                    }
                });
    }

    /*
        İlçe Spinner'ını doldurur.

        District.java içinde toString() name döndürdüğü için
        Spinner ekranda ilçe adını gösterir.
    */
    private void setupDistrictSpinner() {
        ArrayAdapter<District> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                districtList
        );

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        spinnerDistricts.setAdapter(adapter);
    }

    /*
        Şehir + ilçe seçimine göre etkinlikleri backend'den çeker.

        API:
        events/events_by_location.php

        POST:
        api_token
        city_id
        district_id
    */
    private void loadEvents(int cityId, int districtId) {
        setStatus("Etkinlikler yükleniyor...");

        String apiToken = sessionManager.getApiToken();

        ApiClient.getApiService()
                .getEventsByLocation(apiToken, cityId, districtId)
                .enqueue(new Callback<ApiResponse<List<Event>>>() {
                    @Override
                    public void onResponse(
                            Call<ApiResponse<List<Event>>> call,
                            Response<ApiResponse<List<Event>>> response
                    ) {
                        if (!response.isSuccessful()) {
                            setStatus("Etkinlikler alınamadı. Sunucu hatası: " + response.code());
                            return;
                        }

                        ApiResponse<List<Event>> apiResponse = response.body();

                        if (apiResponse == null) {
                            setStatus("Etkinlikler alınamadı. Boş cevap döndü.");
                            return;
                        }

                        if (!apiResponse.isSuccess()) {
                            setStatus(apiResponse.getMessage());
                            return;
                        }

                        List<Event> events = apiResponse.getData();

                        if (events == null) {
                            events = new ArrayList<>();
                        }

                        eventAdapter.updateList(events);

                        if (events.isEmpty()) {
                            setStatus("Bu konum için etkinlik bulunamadı.");
                        } else {
                            setStatus(events.size() + " etkinlik listelendi.");
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<List<Event>>> call, Throwable t) {
                        setStatus("Bağlantı hatası: " + t.getLocalizedMessage());
                    }
                });
    }

    /*
        Durum mesajını ekrana basar.
    */
    private void setStatus(String message) {
        tvStatus.setText(message);
    }

    /*
        Login ekranına döner.
    */
    private void goToLogin() {
        Intent intent = new Intent(HomeActivity.this, LoginActivity.class);
        startActivity(intent);
        finish();
    }
}

File: EBiletKotlin/app/src/main/java/com/alperensarac/ebiletkotlin/ui/home/HomeActivity.kt
package com.alperensarac.ebiletkotlin.ui.home

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.alperensarac.ebiletkotlin.data.api.ApiClient
import com.alperensarac.ebiletkotlin.data.model.ApiResponse
import com.alperensarac.ebiletkotlin.data.model.City
import com.alperensarac.ebiletkotlin.data.model.District
import com.alperensarac.ebiletkotlin.data.model.Event
import com.alperensarac.ebiletkotlin.data.session.SessionManager
import com.alperensarac.ebiletkotlin.databinding.ActivityHomeBinding
import com.alperensarac.ebiletkotlin.ui.auth.LoginActivity
import com.alperensarac.ebiletkotlin.ui.event.EventDetailActivity
import com.alperensarac.ebiletkotlin.ui.scanner.TicketScannerActivity
import com.alperensarac.ebiletkotlin.ui.ticket.MyTicketsActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/*
    HomeActivity

    Ana ekran.

    Görevleri:
    - Kullanıcının giriş yapıp yapmadığını kontrol eder
    - Şehirleri API'den çeker
    - Şehir seçilince ilçeleri API'den çeker
    - İlçe seçilince etkinlikleri listelemeye hazır hale getirir
    - Etkinlikleri RecyclerView içinde gösterir
*/
class HomeActivity : AppCompatActivity() {

    private lateinit var binding: ActivityHomeBinding

    private lateinit var sessionManager: SessionManager

    /*
        Spinner içinde göstereceğimiz şehir ve ilçe listeleri.
    */
    private val cityList = mutableListOf<City>()
    private val districtList = mutableListOf<District>()

    /*
        Seçili şehir ve ilçe.
        API isteği atarken bunların id değerlerini kullanacağız.
    */
    private var selectedCity: City? = null
    private var selectedDistrict: District? = null

    /*
        RecyclerView adapter.
    */
    private lateinit var eventAdapter: EventAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityHomeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        sessionManager = SessionManager(this)

        /*
            Giriş yapılmamışsa LoginActivity'ye gönderiyoruz.
            Çünkü HomeActivity token olmadan çalışamaz.
        */
        if (!sessionManager.isLoggedIn()) {
            goToLogin()
            return
        }

        setupHeader()
        setupRecyclerView()
        setupClickListeners()

        /*
            Ekran açılır açılmaz şehirleri getiriyoruz.
        */
        loadCities()
    }

    /*
        Üstteki kullanıcı bilgisi alanını hazırlar.
    */
    private fun setupHeader() {
        binding.tvWelcome.text = "Hoş geldin, ${sessionManager.getFullName()}"

        val roleText = when (sessionManager.getRole()) {
            "admin" -> "Admin hesabı"
            "staff" -> "Görevli hesabı"
            else -> "Etkinlikleri keşfet"
        }

        binding.tvRoleInfo.text = roleText

        /*
            QR kontrol butonu sadece staff ve admin için görünür.
        */
        binding.btnScanner.visibility = if (sessionManager.isStaffOrAdmin()) {
            View.VISIBLE
        } else {
            View.GONE
        }
    }

    /*
        RecyclerView kurulumu.
    */
    private fun setupRecyclerView() {
        eventAdapter = EventAdapter(
            events = mutableListOf(),
            onEventClick = { event ->

                /*
                    Etkinlik detay ekranına geçiş.

                    Burada tüm event nesnesini göndermiyoruz.
                    Sadece event.id gönderiyoruz.

                    Neden?

                    Çünkü detay ekranında en güncel bilgiyi backend'den çekmek daha doğru.
                    Örneğin kontenjan değişmiş olabilir.
                */
                val intent = Intent(this, EventDetailActivity::class.java)
                intent.putExtra(EventDetailActivity.EXTRA_EVENT_ID, event.id)
                startActivity(intent)
            }
        )

        binding.rvEvents.layoutManager = LinearLayoutManager(this)
        binding.rvEvents.adapter = eventAdapter
    }

    /*
        Buton ve spinner tıklama/ seçim işlemleri.
    */
    private fun setupClickListeners() {
        /*
            Çıkış yapma.
        */
        binding.btnLogout.setOnClickListener {
            sessionManager.logout()
            goToLogin()
        }
        binding.btnScanner.setOnClickListener {
            val intent = Intent(this, TicketScannerActivity::class.java)
            startActivity(intent)
        }
        /*
            Şehir spinner seçimi.

            Kullanıcı şehir seçince ilçeleri getiriyoruz.
        */
        binding.spinnerCities.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(
                    parent: AdapterView<*>?,
                    view: View?,
                    position: Int,
                    id: Long
                ) {
                    if (cityList.isEmpty()) return

                    selectedCity = cityList[position]

                    /*
                        Şehir değiştiğinde eski ilçe ve etkinlik listesini temizliyoruz.
                    */
                    selectedDistrict = null
                    districtList.clear()
                    eventAdapter.updateList(emptyList())

                    loadDistricts(selectedCity!!.id)
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {
                    selectedCity = null
                }
            }

        /*
            İlçe spinner seçimi.
        */
        binding.spinnerDistricts.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(
                    parent: AdapterView<*>?,
                    view: View?,
                    position: Int,
                    id: Long
                ) {
                    if (districtList.isEmpty()) return

                    selectedDistrict = districtList[position]
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {
                    selectedDistrict = null
                }
            }

        /*
            Etkinlikleri listeleme butonu.
        */
        binding.btnListEvents.setOnClickListener {
            val city = selectedCity
            val district = selectedDistrict

            if (city == null) {
                Toast.makeText(this, "Lütfen şehir seçiniz", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (district == null) {
                Toast.makeText(this, "Lütfen ilçe seçiniz", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            loadEvents(city.id, district.id)
        }
        binding.btnMyTickets.setOnClickListener {
            val intent = Intent(this, MyTicketsActivity::class.java)
            startActivity(intent)
        }
    }

    /*
        Şehirleri backend'den getirir.
    */
    private fun loadCities() {
        setStatus("Şehirler yükleniyor...")

        val apiToken = sessionManager.getApiToken()

        ApiClient.apiService.getCities(apiToken)
            .enqueue(object : Callback<ApiResponse<List<City>>> {

                override fun onResponse(
                    call: Call<ApiResponse<List<City>>>,
                    response: Response<ApiResponse<List<City>>>
                ) {
                    if (!response.isSuccessful) {
                        setStatus("Şehirler alınamadı. Sunucu hatası: ${response.code()}")
                        return
                    }

                    val apiResponse = response.body()

                    if (apiResponse == null) {
                        setStatus("Şehirler alınamadı. Boş cevap döndü.")
                        return
                    }

                    if (!apiResponse.success) {
                        setStatus(apiResponse.message)
                        return
                    }

                    val cities = apiResponse.data ?: emptyList()

                    cityList.clear()
                    cityList.addAll(cities)

                    setupCitySpinner()

                    if (cityList.isEmpty()) {
                        setStatus("Aktif şehir bulunamadı.")
                    } else {
                        setStatus("Şehir seçiniz.")
                    }
                }

                override fun onFailure(
                    call: Call<ApiResponse<List<City>>>,
                    t: Throwable
                ) {
                    setStatus("Bağlantı hatası: ${t.localizedMessage}")
                }
            })
    }

    /*
        Şehir spinner'ına şehir adlarını basar.
    */
    private fun setupCitySpinner() {
        val cityNames = cityList.map { it.name }

        val adapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_item,
            cityNames
        )

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

        binding.spinnerCities.adapter = adapter
    }

    /*
        Seçilen şehre göre ilçeleri getirir.
    */
    private fun loadDistricts(cityId: Int) {
        setStatus("İlçeler yükleniyor...")

        val apiToken = sessionManager.getApiToken()

        ApiClient.apiService.getDistrictsByCity(
            apiToken = apiToken,
            cityId = cityId
        ).enqueue(object : Callback<ApiResponse<List<District>>> {

            override fun onResponse(
                call: Call<ApiResponse<List<District>>>,
                response: Response<ApiResponse<List<District>>>
            ) {
                if (!response.isSuccessful) {
                    setStatus("İlçeler alınamadı. Sunucu hatası: ${response.code()}")
                    return
                }

                val apiResponse = response.body()

                if (apiResponse == null) {
                    setStatus("İlçeler alınamadı. Boş cevap döndü.")
                    return
                }

                if (!apiResponse.success) {
                    setStatus(apiResponse.message)
                    return
                }

                val districts = apiResponse.data ?: emptyList()

                districtList.clear()
                districtList.addAll(districts)

                setupDistrictSpinner()

                if (districtList.isEmpty()) {
                    setStatus("Bu şehir için aktif ilçe bulunamadı.")
                } else {
                    setStatus("İlçe seçip etkinlikleri listeleyebilirsin.")
                }
            }

            override fun onFailure(
                call: Call<ApiResponse<List<District>>>,
                t: Throwable
            ) {
                setStatus("Bağlantı hatası: ${t.localizedMessage}")
            }
        })
    }

    /*
        İlçe spinner'ına ilçe adlarını basar.
    */
    private fun setupDistrictSpinner() {
        val districtNames = districtList.map { it.name }

        val adapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_item,
            districtNames
        )

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

        binding.spinnerDistricts.adapter = adapter
    }

    /*
        Seçili şehir + ilçeye göre etkinlikleri getirir.
    */
    private fun loadEvents(cityId: Int, districtId: Int) {
        setStatus("Etkinlikler yükleniyor...")

        val apiToken = sessionManager.getApiToken()

        ApiClient.apiService.getEventsByLocation(
            apiToken = apiToken,
            cityId = cityId,
            districtId = districtId
        ).enqueue(object : Callback<ApiResponse<List<Event>>> {

            override fun onResponse(
                call: Call<ApiResponse<List<Event>>>,
                response: Response<ApiResponse<List<Event>>>
            ) {
                if (!response.isSuccessful) {
                    setStatus("Etkinlikler alınamadı. Sunucu hatası: ${response.code()}")
                    return
                }

                val apiResponse = response.body()

                if (apiResponse == null) {
                    setStatus("Etkinlikler alınamadı. Boş cevap döndü.")
                    return
                }

                if (!apiResponse.success) {
                    setStatus(apiResponse.message)
                    return
                }

                val events = apiResponse.data ?: emptyList()

                eventAdapter.updateList(events)

                if (events.isEmpty()) {
                    setStatus("Bu konum için etkinlik bulunamadı.")
                } else {
                    setStatus("${events.size} etkinlik listelendi.")
                }
            }

            override fun onFailure(
                call: Call<ApiResponse<List<Event>>>,
                t: Throwable
            ) {
                setStatus("Bağlantı hatası: ${t.localizedMessage}")
            }
        })
    }

    /*
        Kullanıcıya ekranda durum mesajı gösterir.
    */
    private fun setStatus(message: String) {
        binding.tvStatus.text = message
    }

    /*
        Login ekranına geçiş.
    */
    private fun goToLogin() {
        val intent = Intent(this, LoginActivity::class.java)
        startActivity(intent)
        finish()
    }
}